### PR TITLE
Spec test utils typescript

### LINKED ISSUE #293: Add spec test utils into lodestar repository
Move https://github.com/ChainSafe/eth2.0-spec-test-util into lodestar repository as lerna package

### LINKED ISSUE #357: Fix test case names
- `operations -  - proposer_slashing operation` looks weird
- distinguish minimal and mainnet spec test cases
- sanity tests are missing block/slot identifier

### LINKED ISSUE #358: [Spec test utils] Rewrite to typescript
Convert this library to typescript

### PR DESCRIPTION
**Pull request by mpetrunic**
I've decided to update incrementally to avoid huge PR so this  is first one moving to typescript with some improvment regarding test names.

- move to typescript
- better types
- improved type generation in build

resolves #358 
resolves #293 
resolves #357

### CODE REVIEW

File: packages/eth2.0-spec-test-util/package.json [hunk 1]
```diff
@@ -0,0 +1,61 @@
+{
+  "name": "@chainsafe/eth2.0-spec-test-util",
+  "version": "0.4.0",
+  "description": "Spec test suite generator from yaml test files",
+  "author": "ChainSafe Systems",
+  "license": "LGPL-3.0",
```
Comment (wemeetagain) [inline packages/eth2.0-spec-test-util/package.json]: can we keep it Apache 2.0?

File: packages/eth2.0-spec-test-util/src/bulk.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+import {describe} from "mocha";
+import {objectToCamelCase} from "./util";
+import {expect} from "chai";
+import profiler from "v8-profiler-next";
+
+export interface BaseCase {
+  description: string;
+}
+
+interface TestSpec<TestCase extends BaseCase> {
```
Comment (wemeetagain) [inline packages/eth2.0-spec-test-util/src/bulk.ts]: Can you add a comment on what TestSpec and BaseCase are?

Eg: TestSpec corresponds to one yaml file of tests.

File: packages/eth2.0-spec-test-util/src/bulk.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  config: string;
+  runner: string;
+  handler: string;
+  testCases: TestCase[];
+}
+
+const env = process.env;
+
+/**
+ * Run yaml Eth2.0 bulk spec tests for a certain function
```
Comment (wemeetagain) [inline packages/eth2.0-spec-test-util/src/bulk.ts]: Bulk vs. ?
I guess this is still built for the old all-in-one-yaml style.

File: packages/ssz/package.json [hunk 4]
```diff
@@ -31,30 +31,12 @@
...
   "license": "Apache-2.0",
   "devDependencies": {
-    "@babel/cli": "^7.2.3",
-    "@babel/core": "^7.3.4",
-    "@babel/plugin-proposal-class-properties": "^7.3.4",
-    "@babel/plugin-proposal-object-rest-spread": "^7.3.4",
-    "@babel/preset-env": "^7.3.4",
-    "@babel/preset-typescript": "^7.3.3",
-    "@babel/register": "^7.0.0",
     "@chainsafe/eth2.0-spec-test-util": "^0.3.0",
```
Comment (wemeetagain) [inline packages/ssz/package.json]: ssz not upgraded? ;)

File: packages/eth2.0-spec-test-util/src/bulk.ts [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          if (env.GEN_PROFILE_DIR) {
+            profiler.startProfiling(profileId);
+          }
+          const result = testFunc(...inputs);
+          if (env.GEN_PROFILE_DIR) {
+            const profile = profiler.stopProfiling(profileId);
+            const directory = env.GEN_PROFILE_DIR || __dirname;
+            profile.export((error, result) => {
+              if (error) {
+                //console.log(error);
```
Comment (wemeetagain) [inline packages/eth2.0-spec-test-util/src/bulk.ts]: remove

**Review by wemeetagain — CHANGES_REQUESTED** [review]

File: packages/eth2.0-spec-test-util/package.json (hunk 1 above)
Comment (mpetrunic) [reply packages/eth2.0-spec-test-util/package.json]: @GregTheGreek ^^

File: packages/eth2.0-spec-test-util/src/bulk.ts (hunk 3 above)
Comment (mpetrunic) [reply packages/eth2.0-spec-test-util/src/bulk.ts]: This is for old.  Any suggestion about naming. file vs dir based?

File: packages/eth2.0-spec-test-util/package.json (hunk 1 above)
Comment (GregTheGreek) [reply packages/eth2.0-spec-test-util/package.json]: Yes - it's fine to make this apache2.0, Lodestar and the lighclient should remain LGPL3.0

File: packages/eth2.0-spec-test-util/src/bulk.ts (hunk 3 above)
Comment (mpetrunic) [reply packages/eth2.0-spec-test-util/src/bulk.ts]: @wemeetagain name suggestion. I'm planning to switch to `describeFileSpec` and `describeDirectorySpec`. Can't think of better naming

Comment (wemeetagain) [reply packages/eth2.0-spec-test-util/src/bulk.ts]: maybe `describeSingleSpec` / `describeMultiSpec` ?

Comment (mpetrunic) [reply packages/eth2.0-spec-test-util/src/bulk.ts]: which one is which? :laughing: I mean only diff is in old one, test cases are in single file and in new one test cases are in directory

Comment (wemeetagain) [reply packages/eth2.0-spec-test-util/src/bulk.ts]: this would be `describeMultiSpec`

**Review by wemeetagain — DISMISSED** [review]

**Review by GregTheGreek — DISMISSED** [review]

**Review by wemeetagain — DISMISSED** [review]

**Review by GregTheGreek — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**